### PR TITLE
Add support for Internal Skin Param

### PR DESCRIPTION
**Pull request by thtrandomlurker**
Adds support for reading, writing, exporting, and importing internal skin param from AC1.01/DT1st

### CODE REVIEW

File: MikuMikuModel/Nodes/Objects/SkinNode.cs [hunk 1]
```diff
@@ -114,6 +119,306 @@ protected override void Initialize()
...
+                                        externalSkinParam.Write("wind_afc", osgBlock.InternalSkinParameter.WindAffection);
+                                    }
+                                    externalSkinParam.PopScope();
+                                }
+                                externalSkinParam.PopScope();
+                            }
+                        }
+                        else if (block is ClothBlock)
+                        {
+                            ClothBlock osgBlock = block as ClothBlock;  // i could've sworn there was a better way to do this
```
Comment (blueskythlikesclouds) [inline MikuMikuModel/Nodes/Objects/SkinNode.cs:201]: See the remark above.

File: MikuMikuModel/Nodes/Objects/SkinNode.cs [hunk 2]
```diff
@@ -114,6 +119,306 @@ protected override void Initialize()
...
+            using (OpenFileDialog dlg = new OpenFileDialog() { Filter = "Skin Parameter (*.txt)|*.txt" })
+            {
+                if (dlg.ShowDialog() == DialogResult.OK)
+                {
+                    ParameterTree externalSkinParam = new ParameterTree(new EndianBinaryReader(File.OpenRead(dlg.FileName), Endianness.Little));
+                    foreach (var block in Data.Blocks)
+                    {
+                        if (block is OsageBlock)
+                        {
+                            var osgBlock = block as OsageBlock;
```
Comment (blueskythlikesclouds) [inline MikuMikuModel/Nodes/Objects/SkinNode.cs:274]: See the remark above.

File: MikuMikuModel/Nodes/Objects/SkinNode.cs [hunk 3]
```diff
@@ -114,6 +119,306 @@ protected override void Initialize()
...
+                                    skp.WindAffection = externalSkinParam.Get<float>("wind_afc");
+                                    osgBlock.InternalSkinParameter = skp;
+                                    externalSkinParam.CloseScope();
+                                }
+                                externalSkinParam.CloseScope();
+                            }
+                        }
+                        else if (block is ClothBlock)
+                        {
+                            var clsBlock = block as ClothBlock;
```
Comment (blueskythlikesclouds) [inline MikuMikuModel/Nodes/Objects/SkinNode.cs:350]: See the remark above.

**Review by blueskythlikesclouds — COMMENTED** [review]
Looks very good, just a few code touchups and it should be done.

File: MikuMikuModel/Nodes/Objects/SkinNode.cs [hunk 4]
```diff
@@ -114,6 +119,306 @@ protected override void Initialize()
...
+            using (SaveFileDialog dlg = new SaveFileDialog() { Filter = "Skin Parameter (*.txt)|*.txt" })
+            {
+                if (dlg.ShowDialog() == DialogResult.OK)
+                {
+                    ParameterTreeWriter externalSkinParam = new ParameterTreeWriter();
+                    foreach (var block in Data.Blocks)
+                    {
+                        if (block is OsageBlock)
+                        {
+                            OsageBlock osgBlock = block as OsageBlock;  // i could've sworn there was a better way to do this
```
[the remark this answers is not in the document]
Comment (thtrandomlurker) [reply MikuMikuModel/Nodes/Objects/SkinNode.cs:134]: I honestly completely forgot this was a thing you could do, i even used this syntax in my own code before, i just forgot about specifying a variable name after the type name, i'll update that right away

**Comment by blueskythlikesclouds** [comment]
Looks good now, do you have anything more to add? If not, I'll merge it.

**Comment by thtrandomlurker** [comment]
Everything should be good now as far as I know, the only thing being that I can't quite understand why collision on skirts specifically don't work correctly, my first guess was that it was due to later games ditching ellipse for sphere, but that didn't seem to affect much. Other collisions import fine though. At any rate, I don't have any more to add, should be good to merge.

**Pull request merged**